core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -318,6 +318,23 @@ boolean isDefaultContext() {
     }
 
     /**
+     * Marks this thread context as an internal system context. This signals that actions in this context are issued
+     * by the system itself rather than by a user action.
+     */
+    public void markAsSystemContext() {
+        if (isSystemContext() == false) {
+            putTransient(INTERNAL_EXECUTION_HEADER, Boolean.TRUE);
+        }
+    }
+
+    /**
+     * Returns <code>true</code> iff this context is a system context
+     */
+    public boolean isSystemContext() {
+        return getTransient(INTERNAL_EXECUTION_HEADER) != null;
```

Comment:
Since putTransient is public, should ensure the key passed in there is never INTERNAL_EXECUTION_HEADER (except in markAsSystemContext above)? Otherwise someone might think they could put false to "override", but then this check would think false is still a system context.